File: talkadvisor-domain/src/test/java/org/craftsrecords/talkadvisor/recommendation/stepdefs/RecommendationStepdefs.java
package org.craftsrecords.talkadvisor.recommendation.stepdefs;

import cucumber.api.java8.En;
import kotlin.ranges.ClosedRange;
import org.craftsrecords.talkadvisor.recommendation.Recommendation;
import org.craftsrecords.talkadvisor.recommendation.api.RecommendTalks;
import org.craftsrecords.talkadvisor.recommendation.criteria.GuestCriteria;
import org.craftsrecords.talkadvisor.recommendation.preferences.Preferences;
import org.craftsrecords.talkadvisor.recommendation.profile.Profile;
import org.craftsrecords.talkadvisor.recommendation.talk.TalkFormat;

import java.time.Duration;

import static java.time.Duration.ofMinutes;
import static kotlin.ranges.RangesKt.rangeTo;
import static org.craftsrecords.talkadvisor.recommendation.assertions.TalkAdvisorAssertions.assertThat;

public class RecommendationStepdefs implements En {
    public RecommendationStepdefs(TestContext testContext,
                                  RecommendTalks recommendTalks) {
        When("^he asks for a recommendation given his criteria$",
                () -> {
                    GuestCriteria guestCriteria = createGuestCriteriaFrom(testContext);
                    Recommendation recommendation = recommendTalks.getRecommendationSatisfying(guestCriteria);
                    testContext.setRecommendationResult(recommendation);
                });
        When("^he asks for a recommendation$", () -> {
            try {
                Recommendation recommendation = recommendTalks.getRecommendationGivenProfile(testContext.getUserId());
                testContext.setRecommendationResult(recommendation);
            } catch (Exception e) {
                testContext.setError(e);
            }
        });

        Then("^talkadvisor recommends some talks$", () -> {
            Recommendation recommendation = testContext.getRecommendationResult();

            assertThat(recommendation).hasTalks();
        });

        Then("^the recommended talks are related to (.+)",
                (String topicName) -> {
                    Recommendation recommendation = testContext.getRecommendationResult();

                    assertThat(recommendation).hasTalksRelatedTo(topicName);
                });

        Then("^all the talks corresponding to the (.+) format have a duration between (\\d+) and (\\d+) minutes$",
                (String format, Integer minDuration, Integer maxDuration) -> {
                    Recommendation recommendation = testContext.getRecommendationResult();
                    TalkFormat talkFormat = TalkFormat.valueOf(format);
                    ClosedRange<Duration> range = rangeTo(ofMinutes(minDuration), ofMinutes(maxDuration));

                    assertThat(recommendation).hasOnlyTalksInTheFormat(talkFormat);
                    assertThat(recommendation).hasOnlyTalksHavingTheirDurationIn(range);
                });

        Then("^the recommended talks correspond to his preferences$", () -> {
            Recommendation recommendation = testContext.getRecommendationResult();
            Profile profile = testContext.getCreatedProfile();
            Preferences preferences = profile.getPreferences();

            assertThat(recommendation).correspondsToTheCriteria(preferences);
            assertThat(recommendation).hasTalksRelatedTo(preferences.getTopics());
            assertThat(recommendation).hasOnlyTalksInTheFormats(preferences.getTalksFormats());
        });
    }

    private GuestCriteria createGuestCriteriaFrom(TestContext testContext) {
        return new GuestCriteria(testContext.getRequestedTopics(), testContext.getRequestedTalksFormats());
    }
}

File: talkadvisor-domain/src/test/kotlin/org/craftsrecords/talkadvisor/recommendation/stepdefs/RecommendationStepdefs.kt
package org.craftsrecords.talkadvisor.recommendation.stepdefs

import cucumber.api.java.en.Then
import cucumber.api.java.en.When
import org.craftsrecords.talkadvisor.recommendation.api.RecommendTalks
import org.craftsrecords.talkadvisor.recommendation.assertions.that
import org.craftsrecords.talkadvisor.recommendation.criteria.GuestCriteria
import org.craftsrecords.talkadvisor.recommendation.talk.TalkFormat
import java.time.Duration.ofMinutes

class RecommendationStepdefs(private val testContext: TestContext,
                             private val recommendTalks: RecommendTalks) {

    @When("^he asks for a recommendation given his criteria$")
    fun `he asks for a recommendation given his criteria`() {
        val guestCriteria = createGuestCriteriaFromContext()

        val recommendation = recommendTalks satisfying guestCriteria

        testContext.recommendation = recommendation
    }

    @When("^he asks for a recommendation$")
    fun `he asks for a recommendation`() {
        try {
            val user = testContext.userId

            val recommendation = recommendTalks to user

            testContext.recommendation = recommendation
        } catch (e: Exception) {
            testContext.error = e
        }
    }

    @Then("^talkadvisor recommends some talks$")
    fun `talkadvisor recommends some talks`() {
        val recommendation = testContext.recommendation

        recommendation.that.hasTalks()
    }

    @Then("^the recommended talks are related to (.+)")
    fun `the recommended talks are related to`(topic: String) {
        val recommendation = testContext.recommendation

        recommendation.that `has talks related to` topic
    }

    @Then("^all the talks corresponding to the (.+) format have a duration between (\\d+) and (\\d+) minutes$")
    fun `all the talks corresponding to the format have a duration in range`(format: String, minDuration: Long, maxDuration: Long) {
        val recommendation = testContext.recommendation
        val talkFormat = TalkFormat.valueOf(format)
        val range = ofMinutes(minDuration).rangeTo(ofMinutes(maxDuration))

        recommendation.that `has only talks in the format` talkFormat
        recommendation.that `has only talks having their duration in` range
    }

    @Then("^the recommended talks correspond to his preferences$")
    fun `the recommended talks correspond to his preferences`() {
        val recommendation = testContext.recommendation
        val profile = testContext.createdProfile
        val preferences = profile.preferences

        recommendation.that `corresponds to the criteria` preferences
        recommendation.that `has talks related to` preferences.topics
        recommendation.that `has only talks in the formats` preferences.talksFormats
    }

    private fun createGuestCriteriaFromContext(): GuestCriteria {
        return GuestCriteria(testContext.requestedTopics, testContext.requestedTalksFormats)
    }
}
